src/construction.cpp

```diff
@@ -2550,42 +2578,82 @@ build_reqs get_build_reqs_for_furn_ter_ids(
-                        ter_id( pre_build.post_terrain ) != base_ter ) ) &&
-                    ( pre_build.pre_terrain.empty() ||
-                      ( pre_build.post_is_furniture &&
-                        ter_id( pre_build.pre_terrain ) == base_ter ) ) &&
-                    pre_build.post_terrain == build_pre_ter &&
-                    pre_build.pre_terrain != build.post_terrain ) {
-                    if( total_builds.find( pre_build.id ) == total_builds.end() ) {
-                        total_builds[pre_build.id] = 0;
+            bool found_pre = false;
+            // only consider DECORATE constructions if there's no other way to build the target
+            // this will allow painting walls, but will skip un-painting walls as a way to make a wall
+            for( bool allow_decorate : {
+                     false, true
+                 } ) {
+                for( const construction &pre_build : constructions ) {
+                    if( ( pre_build.category == construction_category_DECORATE ) != allow_decorate ) {
+                        continue;
+                    }
+                    if( pre_build.category == construction_category_REPAIR ) {
+                        continue;
                     }
-                    total_builds[pre_build.id] += count;
-                    build_pre_ter = pre_build.pre_terrain;
+                    std::string pre_build_pre_terrain = pre_build.pre_terrain.empty() ? "" : *
+                                                        ( pre_build.pre_terrain.begin() );
+                    if( ( pre_build.post_terrain.empty() ||
+                          ( !pre_build.post_is_furniture &&
+                            ter_id( pre_build.post_terrain ) != base_ter ) ) &&
+                        ( pre_build.pre_terrain.empty() ||
+                          ( pre_build.post_is_furniture &&
+                            ter_id( pre_build_pre_terrain ) == base_ter ) ) &&
+                        pre_build.post_terrain == build_pre_ter &&
+                        pre_build_pre_terrain != build.post_terrain ) {
+                        if( pre_build.pre_terrain.size() > 1 ) {
+                            debugmsg( "get_build_reqs_for_furn_ter_ids tried to recurse into %s which has multiple pre_terrain",
+                                      pre_build.str_id.str() );
+                            return;
+                        }
+                        if( total_builds.find( pre_build.id ) == total_builds.end() ) {
+                            total_builds[pre_build.id] = 0;
+                        }
+                        total_builds[pre_build.id] += count;
+                        build_pre_ter = pre_build_pre_terrain;
+                        found_pre = true;
+                        break;
+                    }
+                }
+                if( found_pre ) {
                     break;
                 }
             }
-            break;
```

Comment:
This outermost `break` seems like it made the original algorithm not work. I don't think it ever looked more than one layer deep for blueprint costs. I haven't been able to experimentally confirm, but I wasn't able to see any way it could work with the existing code.